File: Aug/Day20/count-square-submatrices-with-all-ones.py
# Time: O(M * N)  ::  Space: O(M * N)
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        n= len(matrix)
        n1 = len(matrix[0])
        dp = [[0 for _ in range(n1)] for _ in range(n)]
        sumOfSquareSubmatrices = 0
        for i in range(n):
            for j in range(n1):
                if i==0 or j == 0 or matrix[i][j] == 0:
                    dp[i][j] = matrix[i][j]
                else :
                    dp[i][j] = min(dp[i-1][j], dp[i-1][j-1], dp[i][j-1])+1
                sumOfSquareSubmatrices+=dp[i][j]
        return sumOfSquareSubmatrices



# Time: O(M * N)  ::  Space: O(N)
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        n= len(matrix)
        n1 = len(matrix[0])
        prev = [0 for _ in range(n1)]
        sumOfSquareSubmatrices = 0
        for i in range(n):
            curr = [0 for _ in range(n1)]
            for j in range(n1):
                if i==0 or j == 0 or matrix[i][j] == 0:
                    curr[j] = matrix[i][j]
                else :
                    curr[j] = min(prev[j], prev[j-1], curr[j-1])+1
                sumOfSquareSubmatrices+=curr[j]
            prev = curr
        return sumOfSquareSubmatrices

File: Aug/Day20/count-square-submatrices-with-all-ones.java
// Time: O(M * N)  ::  Space: O(M * N)
class Solution {
    public int countSquares(int[][] matrix) {
        int n = matrix.length, n1 = matrix[0].length;
        int[][] dp = new int[n][n1];
        int sumOfSquareSubmatrices = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n1; j++) {
                if (i == 0 || j == 0 || matrix[i][j] == 0) {
                    dp[i][j] = matrix[i][j];
                } else {
                    dp[i][j] = Math.min(Math.min(dp[i - 1][j], dp[i - 1][j - 1]), dp[i][j - 1]) + 1;
                }
                sumOfSquareSubmatrices += dp[i][j];
            }
        }
        return sumOfSquareSubmatrices;
    }
}




//  Time: O(M * N)  ::  Space: O(N)
class Solution {
    public int countSquares(int[][] matrix) {
        int n = matrix.length, n1 = matrix[0].length;
        int[] prev = new int[n1], curr = new int[n1];
        int sumOfSquareSubmatrices = 0;

        for (int i = 0; i < n; i++) {
            curr = new int[n1]; // reset
            for (int j = 0; j < n1; j++) {
                if (i == 0 || j == 0 || matrix[i][j] == 0) {
                    curr[j] = matrix[i][j];
                } else {
                    curr[j] = Math.min(Math.min(prev[j], prev[j - 1]), curr[j - 1]) + 1;
                }
                sumOfSquareSubmatrices += curr[j];
            }
            prev = curr.clone();
        }
        return sumOfSquareSubmatrices;
    }
}
